fix: check TextPart.has_content() before considering it as final result

addypy opened the pull request:
## Description
This PR fixes an issue where streaming responses would stop prematurely when using Ollama as the backend. The problem occurs because Ollama sometimes returns an empty `TextPart` before a `ToolCallPart`, and the agent was treating this empty text as a final result, causing the streaming to stop before processing the tool call.

## Changes
Added a check using `TextPart.has_content()` before considering a text part as a final result:
```python
if isinstance(new_part, _messages.TextPart):
    if new_part.has_content() and _agent_graph.allow_text_result(result_schema):
        return FinalResult(s, None, None)
```

This ensures that:
- Empty text parts are ignored and don't trigger a final result
- Streaming continues properly after receiving empty text parts
- Tool calls are processed correctly even when preceded by empty text parts
- Non-empty text parts still work as expected when they should be final results

## Testing
- Tested with Ollama backend
- Verified that streaming continues properly after tool calls
- Confirmed that non-empty text parts still work as expected for final results
- Tested that the fix doesn't affect other model backends' behavior

## Related Issue
Fixes #1292

## Additional Notes
The issue was particularly noticeable with Ollama's streaming responses, but this fix makes the behavior more robust in general by ensuring we only treat meaningful text content as a final result.

karthik1599 commented:
I'm also facing a similar issue where streaming isn't happening while using Ollama locally. When I checked with `.all_messages()`, the flow stopped after the tool call. It seems like the stream isn't continuing past that point.

samuelcolvin commented on pydantic_ai_slim/pydantic_ai/agent.py:
```diff
@@ -245,7 +245,7 @@ async def run(
         message_history: list[_messages.ModelMessage] | None = None,
         model: models.Model | models.KnownModelName | None = None,
-        deps: AgentDepsT = None,
+        deps: AgentDepsT | None = None,
```
these shouldn't change, since `AgentDepsT` can be `None`.

addypy replied:
While `AgentDepsT` does have `default=None`, removing the explicit `| None` causes these linter errors:

```
Expression of type "None" cannot be assigned to parameter of type "AgentDepsT@Agent"
  Type "None" is incompatible with type "AgentDepsT@Agent"

Argument of type "AgentDepsT@Agent | None" cannot be assigned to parameter "deps" of type "AgentDepsT@Agent"
  Type "AgentDepsT@Agent | None" is incompatible with type "AgentDepsT@Agent"
```

The type system seems to require explicit `| None` annotations to properly type check the flow from public API parameters (where `None` is a default value) through to internal methods.

samuelcolvin commented on pydantic_ai_slim/pydantic_ai/agent.py:
```diff
@@ -1192,7 +1206,7 @@ def _get_model(self, model: models.Model | models.KnownModelName | None) -> mode
         return model_
 
-    def _get_deps(self: Agent[T, ResultDataT], deps: T) -> T:
+    def _get_deps(self: Agent[T, ResultDataT], deps: T | None) -> T:
```
this shouldn't change

addypy replied:
Same issue as above - removing the explicit `| None` causes linter errors:

```
Expression of type "None" cannot be assigned to parameter of type "AgentDepsT@Agent"
  Type "None" is incompatible with type "AgentDepsT@Agent"
```

Again, I think we may need the explicit annotation to handle the `None` default values properly through the call chain.

Kludex replied:
You have an old version of pyright installed.

Also, this project is fully typed with a type checker running in CI, so if it's the way it is, it's because the type checker agrees.

addypy replied:
Thanks @Kludex ! 
I've updated pyright and reverted the type annotation changes, while keeping just the functional fix.

samuelcolvin commented on pydantic_ai_slim/pydantic_ai/agent.py:
```diff
@@ -685,15 +685,29 @@ async def stream_to_final(
                                         for call, _ in result_schema.find_tool([new_part]):
                                             return FinalResult(s, call.tool_name, call.tool_call_id)
+
+                                elif isinstance(maybe_part_event, _messages.PartDeltaEvent) and isinstance(
```
why is this needed?

addypy replied:
You're right - this wasn't needed. I've tested after removing this code and everything works properly. Will revert this change. Thanks for pointing this out!

addypy replied:
Looks like this might be necessary. We need to handle start events and delta events separately. 

Only empty text content in start events should be ignored. while empty text parts in delta events should be allowed to passthrough.

samuelcolvin reviewed (commented):
Thanks so much.

This needs some changes, explanation and definitely tests, but I think it could be a good solution.

addypy commented:
@Kludex @samuelcolvin I've reverted the redundant type annotation changes as requested in the PR. 

I'm seeing test failures that I think might be related to the actual fix for this premature stopping issue (because of the empty TextPart).

The test failures I'm seeing are:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃  File                        ┃  Function                                    ┃  Function Line  ┃  Error Line  ┃  Error           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│  tests/test_streaming.py     │  test_streamed_text_response                 │  37             │  47          │  AgentRunError   │
│  tests/test_streaming.py     │  test_streamed_text_stream                   │  150            │  155         │  AgentRunError   │
│  tests/test_streaming.py     │  test_custom_output_type_default_str         │  773            │  776         │  AgentRunError   │
│  tests/test_streaming.py     │  test_custom_output_type_default_structured  │  785            │  792         │  AgentRunError   │
│  tests/test_usage_limits.py  │  test_streamed_text_limits                   │  69             │  123         │  AssertionError  │
└──────────────────────────────┴──────────────────────────────────────────────┴─────────────────┴──────────────┴──────────────────┘
```

I think these failures might be happening because the fix now checks for `new_part.has_content()` before considering a TextPart as a final result. This appears to be affecting token counting and possibly stream completion logic, as most errors are either `AgentRunError: Should have produced a StreamedRunResult before getting here` or token count assertion errors.

Do you think the test snapshots need to be updated to account for this change in behavior?

addypy commented:
@Kludex  @samuelcolvin 

I’ve cleaned up and removed unnecessary linter issues. The primary work is in  agent.py - in **`stream_to_final`**.

## Key Change
* **State tracking** was added to `stream_to_final` so it can correctly handle mixed output (tool‑calls + text).  
* This prevents empty text parts from ending the stream too early.

---

## Review
1. **State‑tracking approach** in `stream_to_final` – does it look okay?  
2. **Empty text parts** – is the new handling consistent with the expected behaviour?  
3. **Tests** – should any snapshots be updated in light of these changes?

---

<details>
<summary><strong>Summary of Failures (3 failed / 3 errors)</strong></summary>

```text
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃               ┃               ┃  Function     ┃              ┃               ┃
┃  File         ┃  Function     ┃  Line         ┃  Error Line  ┃  Error        ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│  tests/test…  │  test_strea…  │  37           │  67          │  AssertionE…  │
│  tests/test…  │  test_strea…  │  150          │  166         │  AssertionE…  │
│  tests/test…  │  test_strea…  │  69           │  113         │  AssertionE…  │
└───────────────┴───────────────┴───────────────┴──────────────┴───────────────┘
Results (29.38 s):
   3 failed
 996 passed
  37 skipped
   3 errors

make: *** [test] Error 1
```
</details>
